nameko/rpc.py

```diff
@@ -253,149 +318,214 @@ def setup(self):
+    def __init__(self, target_service, **publisher_options):
         self.target_service = target_service
-        self.options = options
+        for option in RESTRICTED_PUBLISHER_OPTIONS:
+            publisher_options.pop(option, None)
+        self.publisher_options = publisher_options
 
-    def get_dependency(self, worker_ctx):
-        return ServiceProxy(
-            worker_ctx,
-            self.target_service,
-            self.rpc_reply_listener,
-            **self.options
-        )
+    @property
+    def amqp_uri(self):
+        return self.container.config[AMQP_URI_CONFIG_KEY]
+
+    def setup(self):
 
+        # ReplyListener.setup() runs concurrently in another thread, so
+        # we need to wait here until it's defined its queue
+        while self.reply_listener.queue is None:
+            time.sleep(.1)
 
-class ServiceProxy(object):
-    def __init__(self, worker_ctx, service_name, reply_listener, **options):
-        self.worker_ctx = worker_ctx
-        self.service_name = service_name
-        self.reply_listener = reply_listener
-        self.options = options
+        exchange = get_rpc_exchange(self.container.config)
 
-    def __getattr__(self, name):
-        return MethodProxy(
-            self.worker_ctx,
-            self.service_name,
-            name,
-            self.reply_listener,
-            **self.options
+        default_serializer = self.container.serializer
+        serializer = self.publisher_options.pop(
+            'serializer', default_serializer
+        )
+
+        self.publisher = self.publisher_cls(
+            self.amqp_uri,
+            exchange=exchange,
+            serializer=serializer,
+            declare=[self.reply_listener.queue],
+            reply_to=self.reply_listener.queue.routing_key,
+            **self.publisher_options
         )
 
+    def get_dependency(self, worker_ctx):
 
-class RpcReply(object):
-    resp_body = None
+        extra_headers = self.get_message_headers(worker_ctx)
 
-    def __init__(self, reply_event):
-        self.reply_event = reply_event
+        publish = partial(self.publisher.publish, extra_headers=extra_headers)
 
-    def result(self):
-        _log.debug('Waiting for RPC reply event %s', self)
+        register_for_reply = self.reply_listener.register_for_reply
 
-        if self.resp_body is None:
-            self.resp_body = self.reply_event.wait()
-            _log.debug('RPC reply event complete %s %s', self, self.resp_body)
+        proxy = Proxy(publish, register_for_reply)
+        return getattr(proxy, self.target_service)
 
-        error = self.resp_body.get('error')
-        if error:
-            raise deserialize(error)
-        return self.resp_body['result']
 
+class Proxy(object):
+    """ Helper object for making RPC calls.
 
-class MethodProxy(HeaderEncoder):
+    The target service and method name may be specified at construction time
+    or by attibute or dict access, for example:
```

Comment:
`attibute/attribute`

Reply:
a bit surprised that the spellchecker (with all its false positives!) isn't catching these...

Reply:
Actually not sure if the sentence is correct. Is `dict access` really another way to specify target service and method name?

Reply:
Yep, you can do `n.rpc["service-name"]["method-name"]()`

Reply:
Spellchecker unfortunately only runs on the `.rst` files, not docstrings. Might be because we don't include the apidoc. I'll see if I can fix it.